go/vt/vtgate/vtgate.go

```diff
@@ -295,8 +295,21 @@ func Init(
 	// TabletGateway can create it's own healthcheck
 	gw := NewTabletGateway(ctx, hc, serv, cell)
 	gw.RegisterStats()
-	if err := gw.WaitForTablets(ctx, tabletTypesToWait); err != nil {
-		log.Fatalf("tabletGateway.WaitForTablets failed: %v", err)
+
+	// Retry loop for potential time-outs waiting for all tablets.
+OuterLoop:
+	for {
+		err := gw.WaitForTablets(ctx, tabletTypesToWait)
+		switch {
+		case err == nil:
+			break OuterLoop
+		case errors.Is(err, context.DeadlineExceeded):
+			log.Warning("TabletGateway timed out waiting for tablets to become available - retrying.")
+
+			continue
```

Comment:
should would require a new context for retry otherwise it will keep failing with `context.DeadlineExceeded`
we should add a test for this change

Reply:
I am unsure I follow what you would like the behavior to be, could you please elaborate? Thanks!

Reply:
We should check the parent context sent to WaitForTablets to ensure it is still valid. If the context has already expired, the retry will continue to fail with the same error.

Reply:
The key decision here is whether we should stop retrying and fail or attempt again with a new context.

Reply:
I see, thanks for the clarification. However, I am unsure this is necessary. The upstream context passed by `Init()`  to `WaitForTablets()` does not seem to be have a timeout; it's [just](https://github.com/ejortegau/vitess/blob/61061cb48480c1700fd0413d5ddec7e5bacf7e18/go/cmd/vtgate/cli/cli.go#L153) a `context.WithCancel()` instead, it is `WaitForTablets()` who creates [a new context with the timeout](https://github.com/ejortegau/vitess/blob/65c00a82ffa567665bda5384ee52ef35818d9833/go/vt/vtgate/tabletgateway.go#L199) given by  `--gateway_initial_tablet_timeout`. Given that, it seems to me that the retries would not fail with the same error - unless the timeout is hit again. Is that not the case?

Reply:
That is my expectation as well.